src/Interpreters/SyntaxAnalyzer.cpp

```diff
@@ -346,6 +346,98 @@ void optimizeGroupBy(ASTSelectQuery * select_query, const NameSet & source_colum
         appendUnusedGroupByColumn(select_query, source_columns);
 }
 
+/// recursive traversal and check for optimizeGroupByFunctionKeys
+int optimizeGroupByFunctionKeysChecks(const ASTPtr& child, std::unordered_map<String, int> * checks)
+{
+    /// check if function has no arguments
+    if ((child->children).empty())
+    {
+        return 1;
+    }
+    for (auto & child_key : child->children)
+    {
+        if (!child_key->as<ASTFunction>())
+        {
+            if ((*checks).find(child_key->getColumnName()) == (*checks).end())
+            {
+                /// if variable of a function is not in GROUP BY keys, this function should not be deleted
+                return 1;
+            }
+        }
+        else
+        {
+            /// if function is not in keys, check for its variables
+            if ((*checks).find(child_key->getColumnName()) == (*checks).end())
+            {
+                return optimizeGroupByFunctionKeysChecks(child_key->as<ASTFunction>()->arguments, checks);
+            }
+        }
+    }
+    return 0;
+}
+
+///eliminate functions of other GROUP BY keys
+void optimizeGroupByFunctionKeys(ASTSelectQuery * select_query)
+{
+    if (!select_query->groupBy())
+    {
+        return;
+    }
+
+    auto & group_keys = select_query->groupBy()->children;
+    auto grp_by = select_query->groupBy();
```

Comment:
It could be not cheap to call the function twice. The best practice is to extract them in another direction:

```
auto grp_by = select_query->groupBy();
auto & group_keys = grp_by->children;
```